Unresolved symbols assertion during pickling when inlining def with result type with non-object prefix

Compiling the following,
```scala
trait Foo {
  type Id[t] = t
  inline def foo[T](t: T) <: Id[T] =
    inline t match {
      case i: Int => (i+1).asInstanceOf[Id[T]]
      case _ => t
    }
}

object Bar extends Foo

object Test {
  Bar.foo(23)
}
```
Results in a compiler crash due to an assertion,
```
exception occurred while compiling tests/pos/inline-from-super.scala
java.lang.AssertionError: assertion failed: unresolved symbols: value Foo_this when pickling tests/pos/inline-from-super.scala while compiling tests/pos/inline-from-super.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: unresolved symbols: value Foo_this when pickling tests/pos/inline-from-super.scala
```
If we move the definition of `Id` out of the definition of `Foo` it compiles,
```scala
object Ext {
  type Id[t] = t
}
import Ext._

trait Foo {
  inline def foo[T](t: T) <: Id[T] =
    inline t match {
      case i: Int => (i+1).asInstanceOf[Id[T]]
      case _ => t
    }
}

object Bar extends Foo

object Test {
  Bar.foo(23)
}
```
It also compiles if we move the entire body of `Foo` into `Bar`,
```scala
object Bar {
  type Id[t] = t
  inline def foo[T](t: T) <: Id[T] =
    inline t match {
      case i: Int => (i+1).asInstanceOf[Id[T]]
      case _ => t
    }
}

object Test {
  Bar.foo(23)
}
```

--- a/compiler/src/dotty/tools/dotc/typer/Inliner.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Inliner.scala
@@ -954,13 +954,18 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(implicit ctx: Context) {
       override def traverse(t: Tree)(implicit ctx: Context) = {
         def updateRefCount(sym: Symbol, inc: Int) =
           for (x <- refCount.get(sym)) refCount(sym) = x + inc
+        def updateTermRefCounts(t: Tree) =
+          t.typeOpt.foreachPart {
+            case ref: TermRef => updateRefCount(ref.symbol, 2) // can't be inlined, so make sure refCount is at least 2
+            case _ =>
+          }
+
         t match {
-          case t: RefTree => updateRefCount(t.symbol, 1)
+          case t: RefTree =>
+            updateRefCount(t.symbol, 1)
+            updateTermRefCounts(t)
           case _: New | _: TypeTree =>
-            t.typeOpt.foreachPart {
-              case ref: TermRef => updateRefCount(ref.symbol, 2) // can't be inlined, so make sure refCount is at least 2
-              case _ =>
-            }
+            updateTermRefCounts(t)
           case _ =>
         }
         traverseChildren(t)

--- a/tests/pos/i5572.scala
+++ b/tests/pos/i5572.scala
@@ -1,0 +1,14 @@
+trait Foo {
+  type Id[t] = t
+  inline def foo[T](t: T) <: Id[T] =
+    inline t match {
+      case i: Int => (i+1).asInstanceOf[Id[T]]
+      case _ => t
+    }
+}
+
+object Bar extends Foo
+
+object Test {
+  Bar.foo(23)
+}

